docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,777 @@
+     |           |                         | ►OPEN     |
+     +-----------+                         +-----------+
+```
+
+
+#### Protocol handler signatures
+
+We first present the signatures of the four protocol handlers; the connection handshake module at each party implements and exposes these handlers.
+
+```golang
+func ConnOpenInit(local ConnectionParty,
+                  remote ConnectionParty)
+
+func ConnOpenTry(local ConnectionParty,
+                 remote ConnectionParty,
+                 remoteVersions string[],
+                 remoteStateProof Proof,
+                 remoteClientProof Proof)
+
+func ConnOpenAck(local ConnectionParty,
+                 remoteVersion string,
+                 remoteStateProof Proof,
+                 remoteClientProof Proof)
+
+func ConnOpenConfirm(local ConnectionParty,
+                     remoteStateProof Proof)
+```
+
+#### Main Protocol Handlers
+
+We first define the four protocol handlers, and then describe the relayer functionality & datagrams.
+Finally, we describe helper functions.
+
+###### ConnOpenInit
+
+This is step 1.
+
+```golang
+func ConnOpenInit(local ConnectionParty, remote ConnectionParty)
+{
+  // Create our own end of the connection.
+  connectionEnd.state = INIT
+  connectionEnd.local = local
+  connectionEnd.remote = remote
+  connectionEnd.version = getCompatibleVersions()
+
+  // Validity check. (Explained below in helper functions).
+  abortTransactionUnless(validate(connectionEnd))
+
+  // Uniqueness check. (Explained below in helper functions).
+  abortTransactionUnless(unique(connectionEnd))
+
+  // Now we save this connectionE in our local store.
+  // Since this connectionEnd is in state INIT, the relayer will continue with
+  // next steps of the connection handshake protocol.
+  setConnection(local.connectionIdentifier, connectionEnd)
+}
+```
+
+Preconditions:
+  - the connection (obtained from the input `local` and `remote` arguments) is valid and unique
+
+Postconditions:
+  - the module executing this handler stores a `ConnectionEnd` (matching with the input arguments `local` and `remote`) in state `INIT`
+  - function produces no output; may abort if preconditions not met
+
+Implementation remarks:
+  - to ensure termination of the ICS3 protocol, this handler must execute at least once (either one of the chains or both)
+
+
+###### ConnOpenTry
+
+This is step 2.b.
+
+```golang
+func ConnOpenTry(local ConnectionParty,
+                 remote ConnectionParty,
+                 remoteVersions string[],
+                 remoteStateProof Proof,
+                 remoteClientProof Proof)
+{
+  // Create our end of the connection.
+  connectionEnd.state = TRYOPEN
+  connectionEnd.local = local
+  connectionEnd.remote = remote
+  connectionEnd.version = pickVersion(remoteVersions)
+
+  // Validity check.
+  abortTransactionUnless(validate(connectionEnd))
+
+  // Uniqueness check.
+  abortTransactionUnless(unique(connectionEnd))
```

Comment:
This should check that either there is no connection of if there is on it is in `INIT` state.